nntrainer/src/neuralnet.cpp

```diff
@@ -124,118 +124,137 @@ int NeuralNetwork::setConfig(std::string config) {
+  popt.decay_rate = decay_rate;
   popt.beta1 = iniparser_getdouble(ini, "Network:beta1", 0.0);
   popt.beta2 = iniparser_getdouble(ini, "Network:beta2", 0.0);
   popt.epsilon = iniparser_getdouble(ini, "Network:epsilon", 0.0);
 
   status = opt.setOptParam(popt);
   NN_INI_RETURN_STATUS();
 
-  for (unsigned int i = 0; i < layers_name.size(); i++)
-    ml_logi("%s", layers_name[i].c_str());
-
-  loss = 100000.0;
-
-  if (iniparser_find_entry(ini, "DataSet:TrainData")) {
-    data_buffer = std::make_shared<DataBufferFromDataFile>();
-
-    std::shared_ptr<DataBufferFromDataFile> dbuffer =
-      std::static_pointer_cast<DataBufferFromDataFile>(data_buffer);
-
-    status = dbuffer->setDataFile(
-      iniparser_getstring(ini, "DataSet:TrainData", ""), DATA_TRAIN);
-    NN_INI_RETURN_STATUS();
-    status = dbuffer->setDataFile(
-      iniparser_getstring(ini, "DataSet:ValidData", ""), DATA_VAL);
-    NN_INI_RETURN_STATUS();
-    status = dbuffer->setDataFile(
-      iniparser_getstring(ini, "DataSet:TestData", ""), DATA_TEST);
-    NN_INI_RETURN_STATUS();
-    status = dbuffer->setDataFile(
-      iniparser_getstring(ini, "DataSet:LabelData", ""), DATA_LABEL);
-    NN_INI_RETURN_STATUS();
-
-  } else if (iniparser_find_entry(ini, "DataSet:Tflite")) {
-    ml_loge("Error: Not yet implemented!");
-    return ML_ERROR_INVALID_PARAMETER;
-  } else {
-    data_buffer = std::make_shared<DataBufferFromCallback>();
+  /** Parse the DataSet section */
+  data_buffer = std::make_shared<DataBufferFromCallback>();
+  section_names_iter =
+    std::find(section_names.begin(), section_names.end(), "dataset");
+  if (section_names_iter != section_names.end()) {
+    section_names.erase(section_names_iter);
+    if (iniparser_find_entry(ini, "DataSet:Tflite")) {
+      ml_loge("Error: Tflite dataset is not yet implemented!");
+      return ML_ERROR_INVALID_PARAMETER;
+    } else {
+      data_buffer = std::make_shared<DataBufferFromDataFile>();
+      std::shared_ptr<DataBufferFromDataFile> dbuffer =
+        std::static_pointer_cast<DataBufferFromDataFile>(data_buffer);
+
+      status = dbuffer->setDataFile(
+          iniparser_getstring(ini, "DataSet:TrainData", ""), DATA_TRAIN);
+      NN_INI_RETURN_STATUS();
+      status = dbuffer->setDataFile(
+          iniparser_getstring(ini, "DataSet:ValidData", ""), DATA_VAL);
+      NN_INI_RETURN_STATUS();
+      status = dbuffer->setDataFile(
+          iniparser_getstring(ini, "DataSet:TestData", ""), DATA_TEST);
+      NN_INI_RETURN_STATUS();
+      status = dbuffer->setDataFile(
+          iniparser_getstring(ini, "DataSet:LabelData", ""), DATA_LABEL);
+      NN_INI_RETURN_STATUS();
+    }
   }
 
+  /** Parse all the layers defined as sections in order */
   TensorDim previous_dim;
-  for (unsigned int i = 0; i < layers_name.size(); i++) {
+  for (section_names_iter = section_names.begin();
+      section_names_iter != section_names.end(); ++section_names_iter) {
     bool last = false;
```

Comment:
So, first layer should go first and last layer should go last, then. Strictly follow in order.

Reply:
Yes, just like the older ini format required `Layers` to give the order of layers.
```
Layers = inputlayer \
         fc1layer \
         batchnormalization \
        outputlayer    #Layers of Neuralnetwork
```

Reply:
Yes, this is the location where the order of layers is defined.
Please refer https://github.com/nnstreamer/nntrainer/pull/240#issuecomment-651039288 for detailed example.